### ISSUE TITLE
Error handling for processes

### ISSUE THREAD

**Issue by micans** [nextflow-io/nextflow#903]
We would like NF to support metadata tracking and routing; this would for example enable a process to inform another process that an action was not taken or a stage not passed, this could trigger other tasks. I include two examples from our current pipeline below; perhaps not a perfect match, but they informed my initial thinking about this.

```groovy
    process irods {
        tag "${samplename}"

        input:
            val samplename from sample_list.flatMap{ it.readLines() }

        output:
            set val(samplename), file('*.cram') optional true into ch_cram_files
            file('*.lostcause.txt') optional true into ch_lostcause_irods

        script:
        """
        if bash -euo pipefail irods.sh -t ${params.studyid} -s ${samplename}; then
          true
        else
          stat=\$?
          tag='UNKNOWN'
          if [[ \$stat == 64 ]]; then tag='nofiles'; fi
          echo -e "${samplename}\\tirods\\t\$tag" > ${samplename}.lostcause.txt
        fi
        """
   }

# later ...
    ch_star_accept = Channel.create()
    ch_star_reject = Channel.create()

    star_aligned
        .choice(ch_star_accept, ch_star_reject)
            { namelogsbams -> check_log(namelogsbams[1]) ? 0 : 1 }

    ch_star_accept
    .map    { name, logs, bams -> [name, bams] }
    .into   { ch_featurecounts; ch_indexbam }

    ch_star_reject
    .map    { it -> "${it[0]}\tSTAR\tlowmapping\n" }
    .mix(ch_lostcause_irods)
    .set    { ch_lostcause }
```

**Comment by micans** [comment]
Additional comment: the IRODS errors I am tracking above may be transient; if I rerun the pipeline hours or a day later this stage may then succeed. However, in this set-up, Nextflow will never retry execution upon resuming the pipeline with `-resume`. If I were to change the code and have non-zero exit status in the failing  `else` branch (where `tag=UNKNOWN`), then my tracking information would not be sent forward. So at the moment my custom meta-data tracking system is not fully functional; either I can have `-resume` working for the irods process or I can have tracking, but I can't have both.

**Comment by pditommaso** [comment]
I think what it could be done is to add some new process event handlers e.g. `onSuccess`, `onError`, `onComplete` and use to track specific tasks events and metadata. For example: 

```
   process irods {
        tag "${samplename}"

        input:
            val samplename from sample_list.flatMap{ it.readLines() }

        output:
            set val(samplename), file('*.cram') optional true into ch_cram_files

        script:
        """
         irods.sh -t ${params.studyid} -s ${samplename}
        """

        onError: 
        if( task.exitStatus==64 ) 
           ch_lostcause_irods << [process: task.process, failed:true, sampleName: sampleName]
        """
   }
```

The idea is to allow process event handlers to access the same process evaluation context including `task` runtime info and input/output files metadata.

**Comment by micans** [comment]
That looks good. In `onError`, would the code still have access to files produced by the `script` section? I don't have a scenario, but perhaps worth considering?

**Comment by micans** [comment]
Just looking at our rnaseq pipeline, I have this tracking in two places, but need it in at least two more places, and generally I would probably use this type of code in the majority of my processes. (see also previous comment).

**Comment by pditommaso** [comment]
> would the code still have access to files produced by the script section? 

It should be possible to access  to output files metadata ie. file names, size, etc.

**Comment by pditommaso** [comment]
Interesting. Yes, the idea would be to expose in the completion handler runtime and input/output files metadata. However in principle it should not allow files I/O manipulation. Do you think this would fit your use case ?

**Comment by micans** [comment]
This is a fairly basic question I fear; with the above code, so
```
      onError: 
        if( task.exitStatus==64 ) 
           ch_lostcause_irods << [process: task.process, failed:true, sampleName: sampleName]
```
How can I later serialise these tuples to a file that can be controlled similar to `publishDir`?  Is this possible in what you propose? I'm still very eager for this feature!

**Comment by pditommaso** [comment]
You could use `ch_lostcause_irods` as any other channel in a downstream process.

**Comment by micans** [comment]
Just to remark that in Simone Coughlan's talk at Nextflow 'Reproducible In silico Genomics' it seemed that she is doing similar accounting tasks in Nextflow code: https://github.com/coughls/nextflow-example-scripts

**Comment by pditommaso** [comment]
I'm thinking that this approach is still too programmatic oriented. At the end the real need here is to capture tasks metadata and save to a file (or eventually to a database, but this could be a sub-scenario). 

It think NF should provide a more declarative approach ie. the process should have a `meta` directive that allows you to declare the metadata attributes you to track. Then the system can collect all these info and save automatically to a file.

**Comment by micans** [comment]
Allright, it sounds like a principled approach. I'll stay tuned to see what happens. This could enable me to create a report as sketched above; but if it is something that happens after workflow completion then I won't be able to send it to MultiQC. I'll try to see if there is something simple that allows me to do this in pure Nextflow; my main current issue is that I cannot have normal resumption and error tracking at the same time. As I see it there are two different issues:

1.  General error/metadata tracking in Nextflow pipelines, for which your proposed `meta` is perfect.
2.  More bespoke tracking for cases of particular interest, where the user will appreciate an application-specific report, such as
    a.  data not ready
    b.  not enough data
    c.  not enough data mapped.

In case (2), there really is interest in a limited set of processes within the pipeline, and it would be nice to manage this in the pipeline by e.g. sending it to something like MultiQC. Do you think the distinction between (1) and (2) is a valid one, or would you say the proposed `meta` will take care of both?

**Comment by pditommaso** [comment]
I would really love to have a better integration with MultiQC. How these info would populate the MultiQC report? Your idea is to create a custom file and add it as an input to the MultiQC step?

**Comment by micans** [comment]
This is working at the moment, as custom content. This is example data that I create using the lostcause channels:
```
# plot_type: 'table'
# section_name: 'Lost samples'
Sample  Process Message
ImmHet6564407 cram  lowreads
ImmHet6564790 cram  lowreads
ImmHet6565818 cram  lowreads
ImmHet6565945 cram  lowreads
```
The comment lines are to instruct MultiQC. This results in this multiqc section:
![lc](https://user-images.githubusercontent.com/3369968/49512610-69416680-f886-11e8-93f8-e4850579f6ee.png)
The work directory and command:
```
farm,0f/10f27f63bd9684fe2aeed79b821e69, ls
farm4b+small_multiqc.html  farm4b+small_multiqc_data  lostcause
farm,0f/10f27f63bd9684fe2aeed79b821e69, cat .command.sh
#!/bin/bash -euo pipefail
multiqc . -f --title "farm4b+small (cellgeni/rnaseq)" --filename "farm4b+small_multiqc.html" -m custom_content -m featureCounts -m star -m fastqc
```
The way I see it now (but happy to learn more) is that MultiQC and nextflow are both very flexible and supply a lot of plumbing options, so I think nothing multiqc-specific is needed in Nextflow. But some generic support in Nextflow could still be useful. I've implemented this for both STAR and hisat2 now, in what seems the best idiom I can come up with:
```
  ch_hisat2_accept = Channel.create()
  ch_hisat2_reject = Channel.create()

  ch_hisat2_aligned
      .choice(ch_hisat2_accept, ch_hisat2_reject)
          { namelogbams -> hisat2_filter(namelogbams[1]) ? 0 : 1 }

  ch_hisat2_accept
  .map    { name, log, bam -> [name, bam] }
  .set    { ch_hisat2_bam }

  ch_hisat2_reject
  .map    { it -> "${it[0]}\thisat2\tlowmapping\n" }
  .mix(ch_lostcause_irods, ch_lostcause_cram, ch_lostcause_star)
  .set    { ch_lostcause }
```
Another pattern I've used twice is this:
```
.... // irods process.
    output:
        set val(samplename), file('*.cram') optional true into ch_cram_files
        file('*.lostcause.txt') optional true into ch_lostcause_irods
... // samtools crams to fastq process (this tests for a minimum number of reads in the script section).
    output:
        set val(samplename), file("${samplename}_?.fastq.gz") optional true into ch_fastqs_irods
        file('*.lostcause.txt') optional true into ch_lostcause_cram
...
```
This one (two channels, optional file names) is actually pretty sweet and short; but for processes where the error is potentially transient (as in the irods process) it stops making the process being activated again under `-resume`. However, this is perhaps an exceptional case; maybe it's actually not that high among my priorities.

**Comment by micans** [comment]
A short update: After the above I've concluded that what I'd like most is to be able to _catch  unforeseen process failures_, e.g. a program crashing or not producing an output unexpectedly, and then _manage that failure as a first-class citizen in the dataflow layer_, e.g. include this in the same 'lost cause' report. I would like to be able to do this without instrumenting every single process. I'm thinking a generic 'onError' with a closure that has a little bit of context (for example the `tag` and exit status, perhaps the input names and `STDERR` output). Perhaps a process could optionally set up some extra custom metadata that can be part of that context.

**Comment by stevekm** [comment]
I am also interested in this, as a method to allow external programs to parse the output of Nextflow pipelines with higher-level logic instead of having to do something like hard-code in the publishDir locations and file naming patterns to look for, which might change. Would be great to have some record of all the meta-data surround each task execution, especially paths to output and publishDir files, along with all the other evaluated directives included in each task.

**Comment by evanbiederstedt** [comment]
Hi @micans @stevekm 

The above would be quite useful. What is the current status with the above? I think the best way forward would be to organize a potential PR, with @pditommaso involved. 

Thanks

**Comment by micans** [comment]
A further thought (experiment). A very general solution in this problem space could be the option to specify that a channel can still be activated when a process fails. This would in one fell swoop allow me to do anything that I want with that error using the full power of Nextflow. The default behaviour would be as it is now; on a task error all its output channels are closed to the task.  One or more output channels could be specified to remain open if an error occurs, perhaps by specifying a closure that for example receives the exit code and perhaps some more context.
Importantly, the task would still be resumed with `-resume`. I like this idea as it is very generic and flexible and does not commit to any framework, hopefully it is not at odds with fundamental design principles.
This would allow various logging methods; for example logging everything, or logging only faulty processes if a channel is for example equipped both with `onError: open` and its inputs as `optional: True`.

**Comment by pditommaso** [comment]
My current idea is to allow the implementation of event handlers on each process scope. For example having 

```
process foo {
  input: .. 
  output: .. 
  "your_command .." 
}
```

Then it should be possible to do: 

```
foo.onError {
  // logic here to handle the error code 
  // `task` implicit variable allow the access to runtime metada, inputs and outputs  
}

foo.onComplete {
  // etc
} 
```

A generic process should be accessible as 

```
process.onError { .. etc  }
```

It should even be possible to pipe the result of event handles to a channel, e.g. 

```
foo.onError { /* logic */  } | error_channel_name
```

**Comment by micans** [comment]
This sounds great Paolo. And this:

```
It should even be possible to pipe the result of event handles to a channel, e.g.

foo.onError { /* logic */  } | error_channel_name
```
Is exactly what I'm hoping for, as it allows a generic user report to be constructed with normal processes, publishDir et cetera, and it will be easy for example to embed it in a multiQC report. The `task` implicit variable would have the metadata and exit code so this creates a lot of power for bespoke reporting.

**Comment by evanbiederstedt** [comment]
Hi @pditommaso 

> My current idea is to allow the implementation of event handlers on each process scope.

I think this makes a great deal of sense, thanks! This should address quite a few concerns. 

Do you have an outline how we the community could best start on this PR? 

Also, I very much agree with @micans here:

> this ... is exactly what I'm hoping for, as it allows a generic user report to be constructed with normal processes, publishDir et cetera, and it will be easy for example to embed it in a multiQC report. The task implicit variable would have the metadata and exit code so this creates a lot of power for bespoke reporting.

Let me know how to best help out---this feature would be invaluable!

**Comment by mes5k** [comment]
YES!  I'd love to see something like this.  I'll echo the call for errors becoming first class citizens in the dataflow so that I can write processes to handle the errors later in my pipeline.  I really want to be able to do something along these lines:

```
nextflow.preview.dsl=2

process foo {
  input:
    val(x)
  output:
    file("${x}.txt")
  script:
  """
  # error prone script to create ${x}.txt
  """
  onError:
    { /* some closure to handle the error */ }
}

foo()

other_process(foo.out)
error_collection_process(foo.err)

```
So, the key ideas are:
- Allow each process to specify its own `onError` closure.
- The closure has access to error metadata as well as input data. Saying that process `foo` failed is useful, but saying that process `foo` failed with input `${x}` is _really_ useful.
- Put the result of the error handler closure into a channel. I'm imagining that every process would have a default `empty` error channel and then only if a handler is provided would the error channel be populated.

To me getting the error into a channel is the most important part so that I can integrate those errors downstream and produce the output I'm required to.

**Comment by micans** [comment]
Hello, there are quite a few ideas on how to approach this. How would you feel @pditommaso  about the following small new feature:

- an output channel can declare e.g. `onFail: true` (similar to `optional: true`).
- if a process fails, only output channels with `onFail: true` are sent to.
- on resume, a failed process will still be re-tried.

This looks to me like a smaller feature, rather than a grand over-arching framework; Perhaps both such a smaller feature and the grand framework could coexist; I'm hoping something like this is small and dedicated and hopefully not too intrusive to implement .

**Comment by mes5k** [comment]
I just created a repo demonstrating a hacked up way to handle errors in _almost_ the way I want things to work.  https://github.com/mes5k/nf_error.  The key requirement for me is to be able to get some sort of error data (including process inputs) into a channel to be used within the pipeline.  In the README for the repo I've described the problems I see with my approach.

**Comment by mes5k** [comment]
@pditommaso if you'd be willing to provide a few pointers on where one might add support for a feature like this, I'd be happy to try and put a PR together.

**Comment by evanbiederstedt** [comment]
I still think this ticket is very much worth pursuing, and I would be happy to help out with a PR. I think we just need some internal agreement/direction upon what precisely needs to be done. :)

**Comment by stevekm** [comment]
Re-reading this ticket, it reminds me of the approach that I demo'd here; https://github.com/stevekm/nextflow-communicator

To manually save meta data to a file in each process then read it with an external service

This would be better handled by the `onComplete` and `onError` handling described here though.

**Comment by pditommaso** [comment]
keep open

**Comment by evanbiederstedt** [comment]
This issue seems valid

**Comment by mes5k** [comment]
Still think this is important!

**Comment by micans** [comment]
Still think this is important.

**Comment by bentsherman** [comment]
We've been discussing this issue and I think we've come up with a good syntax to support this. Should be straightforward to implement as well.

Here's an example:

```groovy
process VALIDATE {
  input:
  path('sample.fastq')

  output:
  path('report.txt')

  error:
  path('error.log')

  script:
  '''
  NUM_LINES=$(wc -l sample.fastq)
  if [[ test -gt $NUM_LINES 100 ]] ; then
    echo "Sample has $NUM_LINES lines." > report.txt
  else
    echo "Sample is invalid, only has $NUM_LINES lines." > error.log
  fi
  '''
}

workflow {
  ch_fastq = channel.fromPath('*.fastq')

  // Alt 1: .out/.err syntax
  VALIDATE(ch_fastq)
  ch_valid = VALIDATE.out
  ch_invalid = VALIDATE.err

  // Alt 2: (out, err) syntax
  (ch_valid, ch_invalid) = VALIDATE(ch_fastq)

  ch_valid.count().view { n -> "Found $n valid samples." }
  ch_invalid.count().view { n -> "Found $n invalid samples." }
}
```

Hope it's clear enough to get the idea across.

* The error condition is triggered when a task execution succeeds but one or more required output files are missing. In this case, the process will try to fulfill the error section and emit values on the error channel instead of the output channel.

* If the script returns a non-zero exit code, or if both output and error sections cannot be fulfilled, the error is handled in the normal way by the error strategy.

* Beyond that, the error channels are handled in the same way as output channels.

**Comment by GallVp** [comment]
Thank you @bentsherman 

To date, we have been using optional outputs for module-level handling of such error. For example, the `FASTAVALIDATOR` module from nf-core/modules uses,

```groovy
    tuple val(meta), path('*.success.log')  , emit: success_log , optional: true
    tuple val(meta), path('*.error.log')    , emit: error_log   , optional: true
    path "versions.yml"                     , emit: versions

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    py_fasta_validator \\
        -f $fasta \\
        2> "${prefix}.error.log" \\
        || echo "Errors from fasta_validate printed to ${prefix}.error.log"


    if [ \$(cat "${prefix}.error.log" | wc -l) -gt 0 ]; then
        echo "Validation failed..."


        cat \\
            "${prefix}.error.log"
    else
        echo "Validation successful..."


        mv \\
            "${prefix}.error.log" \\
            fasta_validate.stderr


        echo "Validation successful..." \\
            > "${prefix}.success.log"
    fi


    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        py_fasta_validator: \$(py_fasta_validator -v | sed 's/.* version //')
    END_VERSIONS
    """
```

The `error` channel does help with making this type of error handling more explicit. However, for some tools like `py_fasta_validator ` which return a non-zero exit code to denote both application errors and system errors, we will still need to use the `statement || statement` logic.

But if we extend the error channel to allow conditional emission, it will give the pipeline developers more control. For example,

```
error:
tuple val(meta), path('application.error.log'), emit: app_errors, codes: [ 1, 2, 4 ]
```

With this sort of logic, the error channel will be emitted even if the module fails with error codes 1, 2 and 4. Thus, the pipeline developer will be responsible for handling these conditions for the module-level errors. For all other error codes, the pipeline-level `errorStrategy` will be triggered.

This pattern is similar to exception handling. Some of the errors can be dealt with at the module-level and the remaining can allowed to pass to the pipeline-level `errorStrategy`.

**Comment by GallVp** [comment]
Another module which has been under review for sometime due to lack of clarity on module-level vs pipeline-level error handling: https://github.com/nf-core/modules/pull/7830#discussion_r2009403837

**Comment by mahesh-panchal** [comment]
One can still test for files or file contents after the `||` to differentiate most times. https://nf-co.re/docs/guidelines/components/modules#capturing-exit-codes

**Comment by mahesh-panchal** [comment]
I have to admit, I feel we already have the tooling to handle within script exit codes by using subshells and emit the exit code in an environment variable.
```bash
( 
set -e
command1
command2
# The exit code of the subshell will be the exit code of the last command executed
)
EXITCODE=$? # Capture the exit code of the subshell
```
The `EXITCODE` can then be sent to a channel since the process succeeds unless it's killed by an out of memory or something.

**Comment by micans** [comment]
I've moved jobs a few years ago, but excited to see this development. The request was to be able to deal with error logic and being able to send (error) files across channels as first-class channel citizens, enabling smoother and more powerful reporting. This fulfills that brief as far as I can see, and I'm not quite sure that other approaches mentioned above have the same ease and effect. The main thing I'm pondering is

> The error condition is triggered when a task execution succeeds but one or more required output files are missing. In this case, the process will try to fulfill the error section and emit values on the error channel instead of the output channel.

An alternative could be to check the error section first - if the error file exists pick the error channels, as the *first* check.
Otherwise, I can envision a case where the process was deficient in some way, produced all regular outputs *and* an error file, and the process task has to remove one of the regular files to trigger the error . I'm not sure about this reasoning one way or another however, the main functionality looks promising.

I assume that
```
error:
```

Can output a tuple of things that can include other files, some or which also exist in the `output:` section.

**Comment by bentsherman** [comment]
> But if we extend the error channel to allow conditional emission, it will give the pipeline developers more control.

The thing about exit codes is that they have no universal meaning, except for one fact -- zero means success, non-zero means failure. For this reason it's better to manage exit codes in the script (as shown by Mahesh) and keep a simple contract between the script and the Nextflow runtime. I believe this is also why the `validExitCodes` directive was removed many years ago.

> An alternative could be to check the error section first - if the error file exists pick the error channels, as the first check.

Good idea, I think this might be better.

> I assume that `error:` can output a tuple of things that can include other files, some or which also exist in the `output:` section.

Yes, the `error:` section should support all the same syntax as `output:`

**Comment by mahesh-panchal** [comment]
I think we should also keep in mind that some tools also write empty error files, and others write errors to stdout along with normal output making testing and capturing issues problematic. I think in practice, there's a high chance that custom code would be needed to make use of this (in which case just use normal outputs).

**Comment by bentsherman** [comment]
I guess you could just delete empty error files in the script to prevent them from triggering the error condition.

For mixed stdout/stderr, you would need some way to detect the error condition in the script and create a separate file for it. For example, you could rename the mixed log to something like `error.log` and match against that filename in the error section.

So I think the bash logic you are already using is part of the solution. The remaining issue is that you have to make all of your output channels optional, to allow for either outputs or errors. But this also allows for both to be missing, which could happen due to some error. Having the separate error section (and channels) solves this issue, so that an output is marked optional only if it is truly optional.

**Comment by micans** [comment]
@mahesh-panchal the error file and error output are under control of the process creator, so I don't see a problem there at all. Of course custom code is needed. I may be missing something ... how would you express Ben's example with normal outputs?

**Comment by mahesh-panchal** [comment]
I would probably implement Ben's example like:
```nextflow
process VALIDATE {
  input:
  path('sample.fastq')

  output:
  path('report.txt'), optional: true, emit: txt
  path('error.log'), optional: true, emit: err

  script:
  '''
  NUM_LINES=$(wc -l < sample.fastq)
  if test "$NUM_LINES" -gt 100; then
    echo "Sample has $NUM_LINES lines." > report.txt
  else
    echo "Sample is invalid, only has $NUM_LINES lines." > error.log
  fi
  test -f report.txt || test -f error.log
  '''
}

workflow {
  ch_fastq = channel.fromPath('*.fastq')

  VALIDATE(ch_fastq)
  ch_valid = VALIDATE.out.txt
  ch_invalid = VALIDATE.out.err

  ch_valid.count().view { n -> "Found $n valid samples." }
  ch_invalid.count().view { n -> "Found $n invalid samples." }
}
```

> But this also allows for both to be missing, which could happen due to some error.

Using 
```bash
test -f report.txt || test -f error.log
```
is enough to make sure the correct files are present, when they are optional.

**Comment by micans** [comment]
Thanks - Ben raised

> But this also allows for both to be missing, which could happen due to some error. Having the separate error section (and channels) solves this issue, so that an output is marked optional only if it is truly optional.

Maybe it's a case of how the benefits are weighed against the costs.
@bentsherman at the start of my request I mentioned the issue of `-resume`. In this proposal, would a task always be resumed if the `error:` section was triggered? I'm a bit conflicted. In the IRODS example I gave this would be useful. In the 'insufficient reads' example it would not. Actually (thinking out loud)

- for the IRODS example a previously activated `error:` section with `-resume` triggering the task to run again makes sense.
- for the 'insufficient reads' example Mahesh-Panchal's example makes sense (`-resume` not triggering).

So `error:` would be useful for reporting on conditions that one expects to be transient. I imagine that may be too niche, but in the original scenario when I opened this issue it would be very useful.

**Comment by bentsherman** [comment]
> In this proposal, would a task always be resumed if the `error:` section was triggered?

That's a great question. Thinking out loud...

I think of these errors as two categories, "domain" errors and "execution" errors. A domain error will always happen because the input is invalid (e.g. insufficient reads), whereas an execution error could be avoided given the right configuration, compute resources, etc.

So I think the `error:` section is intended for domain errors while `errorStrategy` is designed for execution errors. Based on that dichotomy, execution errors should always be re-executed on a resume whereas domain errors should be cached -- and you should use `error:` and `errorStrategy` accordingly.

At the same time, I appreciate that, as a practical matter, you might want to treat certain execution errors (e.g. insufficient memory) as a domain error (i.e. a "lost cause") so that you can proceed with the pipeline.

This is essentially what the `ignore` error strategy is for, it just doesn't give you a nice summary of ignored tasks and why they failed. Maybe `ignore` could be improved to close this gap -- I just put down some ideas for this in #6155

Another way might be to add a new error strategy:
```groovy
errorStrategy { task.attempt < 3 ? 'retry' : 'emit' }
```

This would try to emit the error using the `error:` section instead of terminating the pipeline. Of course you can't really guarantee the presence of any output / error files in this case, so your error section would need to be flexible (i.e. optional files).

To return to your question about resume -- this way, you could pass off an execution error as a domain error, record the error e.g. in some report, and on a resume, Nextflow would cache the domain error instead of re-executing. That saves you from executing a few tasks that you know are doomed to fail anyway, which I think is what you were getting at

**Comment by micans** [comment]
In this terminology, it seems that the domain error can be handled in the way of Mahesh-Panchal's example. When I opened the issue, I was thinking about better reporting for execution errors (such as transient IRODS availability issues) and being able to create my own reporting with all NF capabilities available. (I see in my example I mixed the domain error channel (low reads) with the execution error channel. This was at a time when I ran projects containing tens of thousands of samples). Right now (current job) these issues are not a concern so I have no pressing need now and sadly would not be first in line to test.
The functionality could apply much broader than IRODS, including configuration/resources as you mention. IRODS is interesting as an example of a wider class of unknown issues that one may yet encounter.

In summary, the essence of the request (from my pov) was
- on error condition allow (if desired) custom reporting on user side using all NF capabilities
- resumption is still triggered for these processes (this is pretty key).

This opens up powerful error reporting. I think I'd prefer it if this does not depend on the `errorStrategy`, that is, in the absence of any `errorStrategy` the error channel would activate (if conditions are fulfilled).
From the responses here, it seems that domain errors is what most people are dealing with, rather than execution errors. Maybe there is not sufficient demand then.

**Comment by mahesh-panchal** [comment]
I think there's still a lot of demand for better error handling in processes. A blog post and/or implementation patterns can help with the situation here so people become more familiar with the tools available to them. This is also bash specific. I'm not sure what the corresponding patterns would be if it was an R or Python script instead for example, although there should be something similar. 

I feel like the equivalent in another scripting language might be a bit more involved, so this might still be valuable.

**Comment by muffato** [comment]
I really haven't thought it through, so don't hesitate to shoot my idea down.

I see a similarity with try-catch blocks, and I find bash implementations may require advanced bash knowledge. What about an `error:` block that would only run after `script:` if the script has a non-zero return code ?
There would need to be a convention for:

- inheriting all the variables from the `script:` block into `error:`,
- adding the return code as a parameter (e.g. `task.exit`),
- the `error:` block declaring whether the error is recoverable or fatal (maybe just a bash `exit` statement at the end, with the usual zero vs non-zero convention)

And I'd use the same output channels as already defined in `output:`, still making some optional like @GallVp wrote in his comment https://github.com/nextflow-io/nextflow/issues/903#issuecomment-2932867133

**Comment by mahesh-panchal** [comment]
What about an extra flag on the `output:` block instead?

```
  output:
  path('report.txt'), emit: txt, onError: false
  path('error.log'),  emit: err, onError: true
```

With the meaning that:
- without the flag: it's emitted as standard
- `onError: false`: required only if the script exits successfully, and is emitted on success 
- `onError: true`: required only if the script exits with non-zero, and is emitted on failure. 

This comes with the implication though that a process can emit half-baked results on error, which probably means the majority of output channels probably want the `onError: false` flag. Hmmm. Maybe not such a good idea then.

Maybe then:
- emitIf: 'success' ( default )
- emitIf: 'always'
- emitIf: 'error'

**Comment by bentsherman** [comment]
> An alternative could be to check the error section first - if the error file exists pick the error channels, as the first check.

On second thought, this might not work. Imagine you have an error section that simply emits the inputs:

```groovy
process fastqc {
  input:
  tuple val(id), path('sample.fastq')

  output:
  tuple val(id), path('fastqc_logs')

  error:
  tuple val(id), path('sample.fastq')

  script:
  // ...
}
```

This seems like a reasonable use case, but it only works if we try to fulfill the output section first.

In the event of a domain error where the outputs happen to be present (but perhaps are incomplete / invalid), the script should delete these files to ensure that the error condition is triggered.

> This is also bash specific. I'm not sure what the corresponding patterns would be if it was an R or Python script instead for example, although there should be something similar.

You would use the error handling facilities of those languages, e.g. try/catch in Python.

The contract that Nextflow would offer you is this -- to raise a domain error, return a zero exit code and make sure that the output section can't be fulfilled, then I will emit the error section instead. Alternatively, use the `emit` error strategy. This contract is agnostic to the scripting language that you use.

> What about an `error:` block that would only run after `script:` if the script has a non-zero return code ?

A few people have suggested using the exit code. This was also my original feeling, but I think it might be the wrong approach. An exit code doesn't really give you enough context to make an informed decision:

* which command did it come from?
* what does it actually mean in environment X vs Y vs Z?

This is why I think we need to use the presence/absence of outputs to decide whether to emit a domain error. Think of the error section as an alternative output -- try to emit the output, and failing that emit the error.

**Comment by mahesh-panchal** [comment]
> An exit code doesn't really give you enough context to make an informed decision:
>
> -  which command did it come from?


At least for bash scripts you can use:
```
trap 'echo "Script exited on line ${LINENO} with exit status $?" >&2' ERR
```
maybe in the .command.run or something, but there's some incompatibility with `set -e` since that exits immediately so you need to set `set -o errtrace` or `set -E` or something. I don't recall exactly.

**Comment by bentsherman** [comment]
But the Nextflow runtime doesn't know about that message, it only receives the exit code. I suppose it could check the task error log, but then it has to know how to parse it without picking up false positives, and the user has to format the message exactly the right way... that doesn't seem like a good interface either

**Comment by michaeljon** [comment]
It would be really helpful to have a first-class error path back to the nextflow runtime that could be caught by any shell running that runtime. I like the idea of separating these into domain failure cases vs. infrastructure / runtime failure cases. The latter we can do something about using retry strategies or by tweaking our AWS infra (these usually point out something we might have wrong in our batch configuration). The former are the ones were we want to pretend that the pipeline succeeded, but be able to have the nextflow process exit with some other error than 0 or 1.

**Comment by mahesh-panchal** [comment]
You can also run
```
nextflow run ... || test -f run_failed.txt
```
Where `run_failed.txt` can be a file written in `workflow.onError` based on some condition. You can also use a native process to write out a file where `nextflow` is run too, if you need the output to depend on channel content.

Although I do agree it would be nice to be able to exit with different codes and messages and still run the nextflow cleanup steps properly.

**Comment by bentsherman** [comment]
Another interesting consideration here: #6458

When a process is called with only value channels, the output will also be a value channel. The general assumption here is that a value channel should always have a value, since it's not a collection like a queue channel.

So, if I call a process with a value channel and I have both `output:` and `error:` sections, should it return (1) two separate value channels, where one will be empty (or perhaps have a `null` value) or (2) one value channel, which contains the output value or error value depending on the outcome

```groovy
nextflow.preview.types = true

process fastqc {
  input:
  (id, fastq): Tuple<String,Path>

  output:
  tuple(id, file('fastqc_logs'))

  error:
  tuple(id, fastq)

  // ...
}

workflow {
  // separate output and error channels
  (out, err) = fastqc( tuple('1', file('1.fastq')) )

  // single channel with tuple of (output. error)
  fastqc = fastqc( tuple('1', file('1.fastq')) )
  fastqc.subscribe { out, err ->
    // ...
  }
}
```

I guess this question applies to both queue channels and value channels. Something to think about

**Comment by Hoohm** [comment]
So I've been trying to tackle the same issue in our pipeline. Example: "Not enough reads" or "Not enough cells".

I've been using optional outputs to enable/disable downstream processes that depend on having reads or cells. I also added an optional "status" topic output that reports these events so I can know what happened.

This works very well for:
1. Always finish the run, no errors
2. Dynamically enable/disable per sample pipeline wide
3. Allows for neat behaviors such as never checking if the inputs are empty

This does not work well for :
1. Reporting at the end of the run what "happened"
2. Which process for which sample did not produce a file because "not enough reads"/"Not enough cells"

I've been using the same approach (conditional pathing) for missing input files (such as references) to enable/disable whole workflows that require does using `join` which becomes "empty" when does entries are missing.

But I have the same issues as reported just before.

This has pushed me into unholy things such as making my own DAG, tracking "errors" as an optional topic that I collect at the end and write history as the last step of the pipeline. It kind of works, but it's messy, lot of boilerplate code for something that probably should be done differently.

Here is an example of nf-metro output of a sample were cell calling failed on the amplicon path:

<img width="1803" height="750" alt="Image" src="https://github.com/user-attachments/assets/e2ac376e-33d8-4114-bcfc-e1e3480b2026" />

This was done manually as the feature it not totally done. But the backend has the following:
1. Custom manually created "generalistic" DAG
2. Custom python code that takes pre-runtime information such as absence/presence of refs, runtime events that disable downstream processes (still struggling with the refs presence/absence catching, but it's on the way.
3. "Smart annotation of the DAG based on 2
4. Generation of the nf-metro plot in a process
5. Integration into multiqc

I would love to not have to implement point N°2 or at least par of it.

Now to your proposal.

Two output channels (out/err), one "empty" one "filled". I would change:
1. Delete my status topic and use the `error` channels instead to track what happened.

One output channel with both outputs. Would this mean I could access them like this:
`PROCESS.out.foo.output` `PROCESS.out.foo.error`?

Both of these could work, but they solve the runtime DAG generation that I can use to "know" what happened?

**Comment by bentsherman** [comment]
@Hoohm with records, we are moving towards a pattern of one "fat record" output instead of multiple skinny outputs. So with this feature you would simply have an `output:` record and an `error:` record, and you would access them like so:

```
(out, err) = PROCESS(...)
out.view()
err.view()
```
